proposals/NNNN-byte-address-buffer-alignment.md

```diff
@@ -0,0 +1,983 @@
+```
+
+> **Implementation note**: The `BaseAlignment` attribute affects both the base address alignment information in DXIL
+> resource properties and the operation-level alignment calculation for existing `Load`/`Store` functions. This ensures
+> consistent alignment semantics across all buffer access methods. Code that doesn't use `BaseAlignment` continues to
+> work unchanged. Code that adds `BaseAlignment` may experience improved alignment for current operations.
+
+#### Behavior of AlignedLoad/AlignedStore Functions without BaseAlignment Attribute
+
+When the `BaseAlignment` attribute is _not_ specified on the buffer, but the new `AlignedLoad`/`AlignedStore` functions
+are used, then DXC uses the existing alignment requirements to calculate the effective alignment:
+
+> **HLSL Specification, 1.7.2 Memory Spaces**: _"The alignment requirements of an offset into device memory space is the
+> size in bytes of the largest scalar type contained in the given aggregate type"_
+
+If the `alignment` parameter passed to the `AlignedLoad`/`AlignedStore` functions is smaller than the scalar alignment,
+then the compiler will issue an error message.  If the value is larger, then the compiler will issue a warning message
+to inform the developer that this mismatch may be unexpected.  However, for the sake of code reuse, this mismatch will
+be allowed.
+
+#### BaseAlignment Attribute on Function Parameters
+
+The `BaseAlignment` attribute can be applied to `[RW]ByteAddressBuffer` function parameters to specify alignment
+requirements during parameter passing. This enables functions to declare their alignment assumptions explicitly and
+ensures that callers provide buffers with sufficient alignment guarantees. The attribute follows standard alignment
+decay rules where parameter alignment can be less than or equal to the argument's declared alignment, but cannot exceed
+it.
+
+When a function parameter specifies `BaseAlignment`, the argument passed to that function must have been declared with
+`BaseAlignment`, and the argument's alignment value must be greater than or equal to the parameter's alignment
+requirement. If the parameter does not specify `BaseAlignment`, arguments with or without `BaseAlignment` can be passed.
+The effective alignment calculation within the function uses the parameter's declared `BaseAlignment` value, not the
+argument's original alignment, ensuring consistent behavior regardless of the caller's buffer alignment.
+
+Function parameters with `BaseAlignment` maintain the same alignment semantics as buffer declarations: they affect both
+the base address alignment information passed to DXIL operations and the effective alignment calculations for
+`Load`/`Store` and `AlignedLoad`/`AlignedStore` functions called within the function scope. This allows functions to be
+written with specific alignment assumptions while maintaining type safety and preventing alignment-related errors at
+compile time.
```

Comment:
I think that base address alignment passed to DXIL `AnnotateHandle` should ultimately be the root alignment when `AnnotateHandle` chains are collapsed, as they may be from inlining and other optimizations. This means the attribute on the parameter may be used to verify that it is within the base alignment of a passed argument in Sema, but may be dropped in DXIL in favor of the original buffer's base alignment (which would be greater than or equal to the parameter's alignment).